source/common/common/utility.cc

```diff
@@ -20,8 +22,105 @@
+
+  const std::chrono::seconds epoch_time_seconds =
+      std::chrono::duration_cast<std::chrono::seconds>(epoch_time_ns);
+
+  if (cached_time.formatted.find(format_string_) == cached_time.formatted.end() ||
+      cached_time.epoch_time_seconds != epoch_time_seconds) {
+    cached_time.formatted.emplace(
+        std::make_pair(format_string_, fromTime(std::chrono::system_clock::to_time_t(time))));
+
+    cached_time.epoch_time_seconds = epoch_time_seconds;
+  }
+
+  if (subseconds_.empty()) {
+    return cached_time.formatted.at(format_string_);
+  }
+
+  // Copy the cached formatted string, then replace its subseconds part.
+  std::string formatted = cached_time.formatted.at(format_string_);
+  const std::string value = fmt::FormatInt(epoch_time_ns.count()).str();
+  for (const auto subsecond : subseconds_) {
+    // TODO(dio): Infer the length of second from parsing step. Currently, it is defaulted to 10.
+    formatted.replace(subsecond.position_, subsecond.width_, value.substr(10, subsecond.width_));
+  }
+
+  // TODO(dio): Assert the formatted length.
+  return formatted;
+}
+
+std::string DateFormatter::parse(const std::string& format_string) {
+  const auto& specifier_begin = std::sregex_iterator(format_string.begin(), format_string.end(),
+                                                     SubsecondConstants::get().PATTERN);
+  const auto& specifier_end = std::sregex_iterator();
+  if (specifier_begin == specifier_end) {
+    return format_string;
+  }
+
+  std::string new_format_string = EMPTY_STRING;
+  subseconds_.reserve(std::distance(specifier_begin, specifier_end));
+
+  auto now = std::chrono::system_clock::now();
+  time_t current_time = std::chrono::system_clock::to_time_t(now);
+  tm current_tm;
+  gmtime_r(&current_time, &current_tm);
+
+  size_t start = 0;
+  int32_t delta = 0;
+  int32_t inner_delta = 0;
+  std::array<char, 1024> buf;
+  for (std::sregex_iterator i = specifier_begin; i != specifier_end; ++i) {
+    const std::smatch& match = *i;
+
+    const std::string piece = format_string.substr(start, match.position() - start);
+    const size_t formatted_length = strftime(&buf[0], buf.size(), piece.c_str(), &current_tm);
+    delta += static_cast<int32_t>(formatted_length - piece.size()) + inner_delta;
+    start = match.position() + match.length();
+
+    // Save a subsecond's position and width inside the rendered format string.
+    SubsecondSpecifier subsecond(match.position() + delta);
+    const std::string match_str = match[1].str();
+    if (!match_str.empty()) {
+      subsecond.width_ = match_str.at(0) - '0';
+    }
+    subseconds_.emplace_back(subsecond);
+
+    absl::StrAppend(&new_format_string, piece,
+                    SubsecondConstants::get().PLACEHOLDER.substr(0, subsecond.width_));
+
+    // This takes into account the difference between the specifier length and the desired
+    // rendered width.
+    inner_delta = static_cast<int32_t>(subsecond.width_ - match.length());
```

Comment:
This seems more complicated than necessary, any way to simplify it down? I think it might be fine to do something really dumb like:
1. Find the first %f pattern.
2. Record the particular (position, width).
3. Replace in-place the %f with the PLACEHOLDER.
4. Repeat 1.
This is pretty bad in the worst case since it is O(n^2) in string length, but in reality you probably only have a single %f pattern, so it's just as good as the clever way of dealing with it. I think you could write this in about 5 lines of code though :)

Reply:
OK, I will try.